pints/_optimisers/_adam.py

```diff
@@ -0,0 +1,163 @@
+
+        # Best solution found
+        self._x_best = self._x0
+        self._f_best = float('inf')
+
+        # Current point, score, and gradient
+        self._current = self._x0
+        self._current_f = np.inf
+        self._current_df = None
+
+        # Proposed next point (read-only, so can be passed to user)
+        self._proposed = self._x0
+        self._proposed.setflags(write=False)
+
+        # Moment vectors
+        self._m = np.zeros(self._x0.shape)
+        self._v = np.zeros(self._x0.shape)
+
+        # Powers of decay rates
+        self._b1t = None
+        self._b2t = None
+
+        # Exponential decay rates for the moment estimates
+        self._b1 = 0.9    # 0 < b1 <= 1
+        self._b2 = 0.999  # 0 < b2 <= 1
+
+        # Step size
+        self._alpha = np.min(self._sigma0)
+
+        # Small number added to avoid divide-by-zero
+        self._eps = 1e-8
+
+    def ask(self):
+        """ See :meth:`Optimiser.ask()`. """
+
+        # Running, and ready for tell now
+        self._ready_for_tell = True
+        self._running = True
+
+        # Return proposed points (just the one)
+        return [self._proposed]
+
+    def f_best(self):
+        """ See :meth:`Optimiser.f_best()`. """
+        return self._f_best
+
+    def f_guessed(self):
+        """ See :meth:`Optimiser.f_guessed()`. """
+        return self._current_f
+
+    def name(self):
+        """ See :meth:`Optimiser.name()`. """
+        return 'Adam'
+
+    def needs_sensitivities(self):
+        """ See :meth:`Optimiser.needs_sensitivities()`. """
+        return True
+
+    def n_hyper_parameters(self):
+        """ See :meth:`pints.TunableMethod.n_hyper_parameters()`. """
+        return 0
+
+    def running(self):
+        """ See :meth:`Optimiser.running()`. """
+        return self._running
+
+    def tell(self, reply):
+        """ See :meth:`Optimiser.tell()`. """
+
+        # Check ask-tell pattern
+        if not self._ready_for_tell:
+            raise Exception('ask() not called before tell()')
+        self._ready_for_tell = False
+
+        # Unpack reply
+        fx, dfx = reply[0]
+
+        # First iteration
+        if self._current_df is None:
+            self._current_f = fx
+            self._current_df = dfx
+            self._f_best = fx
+            self._x_best = self._current
+
+            # Set bxt to bx^1 for 1st "proper" iteration
+            self._b1t = self._b1
+            self._b2t = self._b2
+            return
+
+        # Update current point
+        self._current = self._proposed
+        self._current_f = fx
+        self._current_df = dfx
+
+        # "Update biased first moment estimate"
+        self._m = self._b1 * self._m + (1 - self._b1) * dfx
+
+        # "Update biased secon raw moment estimate"
+        self._v = self._b2 * self._v + (1 - self._b2) * dfx**2
+
+        # "Compute bias-corrected first moment estimate"
+        m = self._m / (1 - self._b1t)
+        self._b1t *= self._b1
```

Comment:
Can we add a comment here that we update b1 and b2 at the end as opposed to the start of the loop because of the initialisation loop in line 117? 

I think this implementation is equivalent to the pseudo code in the reference that updates t in the beginning, but it might not be immediately obvious.

Reply:
I wrote that in a slightly silly way. Have updated now and skipped the unneeded first iteration